Start integration of SOS

tom111 opened the pull request:
Add SemidefiniteProgramming Package
Add SOS Package (including cached example outputs)

tom111 commented:
Hi Mike, this is the code for SDP and SOS.

- At the moment, SOS has cached example outputs that we might remove later
- I've not configured the path to the included executable correctly. Please do that in the SDP package.

tom111 commented:
Hi again. I'm now adding Dan also as assigned (In the hope it is not too pushy). I think the top level code for SOS and SDP is in good shape now. Mike's makefile in the BUILD directory contains everything that is needed to build CSDP with gcc8 on the mac. Before the merge of this, somebody needs to setup build of csdp on all supported platforms/compilers. Let me know if there is something I can help with!
Thanks.

tom111 commented:
OK, so what needs to be done here? All platforms have a compiler that works. GCC is the default on Linux and on MacOS there is gcc or clang from homebrew (in which case -L flags and other things might be necessary).

We need some logic in the build system to figure out in which compiler situation we are and what the appropriate flags are, right?

DanGrayson commented:
No, I think it is up to the builder of Macaulay2 to specify the compiler and its corresponding -L and -I flags.  But if we start requiring the compiler to accept the -fopenmp flag, then we should have a test in configure.ac to assure that it does.

mikestillman commented:
We can put our recommendations about CC, CXX on the mac into the INSTALL file.

I don’t know how to write configure.ac tests.  Dan, is there one that you have written that I can crib off of?

tom111 commented:
In this manual:
https://www.gnu.org/software/autoconf-archive/ax_openmp.html
it seems to say: 

- If autoconf is 2.62 or newer, there is a macro `ax_openmp` which figures out if openmp programs can be compiled and if so, populates the variables OPENMP_CFLAGS/OPENMP_CXXFLAGS/OPENMP_F77FLAGS which we need to add to our compile and link calls.

Is that it?

DanGrayson commented:
That sounds good.

On Thu, Oct 11, 2018 at 2:15 PM Thomas Kahle <notifications@github.com>
wrote:

> In this manual:
> https://www.gnu.org/software/autoconf-archive/ax_openmp.html
> it seems to say:
>
>    - If autoconf is 2.62 or newer, there is a macro ax_openmp which
>    figures out if openmp programs can be compiled and if so, populates the
>    variables OPENMP_CFLAGS/OPENMP_CXXFLAGS/OPENMP_F77FLAGS which we need to
>    add to our compile and link calls.
>
> Is that it?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Macaulay2/M2/pull/850#issuecomment-428932849>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAqvRB9PzQK3VathlsOd88tW8Fg2tsi6ks5ujzZ3gaJpZM4Wha5D>
> .
>

DanGrayson commented on M2/configure.ac:
```diff
@@ -740,6 +740,28 @@ AC_COMPILE_IFELSE([AC_LANG_SOURCE([
+  AC_SUBST(OPENMP_CXXFLAGS,"$OPENMP_CXXFLAGS")
+  AC_SUBST(OPENMP_LDFLAGS,"")
+  fi
+else
```
What's the point of giving the user the option of making us not test whether openmp works?  It's better to always test it.  Also, what's the point of OPENMP_LDFLAGS?  Just let the user use LDFLAGS

mikestillman replied:
The problem is that under clang, the macro fails, as far as I can tell.  I don't really want to rewrite the macro.  It certainly fails under apple clang, and I think that even under clang-6, one needs to add `-fopenmp -lomp` for the linker.  Clang also warns if it has flags it doesn't understand on the command line.

To get OpenMP to work under apple clang, one needs to first
  `brew install llvm`
which gives one access to libomp.a. Of course, after installing llvm, one can use a later, probably better, clang, since llvm includes clang-7 (currently).

After installing that, add to CFLAGS or CXXFLAGS or CPPFLAGS (not sure where it really belongs):
`-I/usr/local/opt/libomp/include -Xpreprocessor -fopenmp`
and add to LDFLAGS, LDLIBS, whichever is better:
`-L/usr/local/opt/libomp/lib -lomp`
(and also the `-Xpreprocessor -fopenmp` should also be there for linking, I think. It is for me, as the CXXFLAGS get used on the linker command.

For latest clang, AC_OPENMP realizes it needs -fopenmp, but doesn't realize one needs to still add -lomp.  Actually, I need to check this case again.  I've been trying too many variations...

I'm not sure the best way to get this to go.  I already have enough trouble doing autoconf stuff...!  I guess we can add -lomp to LDFLAGS on the configure line, and the other stuff too. And change the compiler to include the needed flags above.  Is that better?  Then maybe AC_OPENMP will work on all cases.  But then we will effectively be using openmp in our code, which we don't (currently) need.  

I'll try this last solution, and see if it works under the 3 cases: apple clang, latest clang (6 or 7), and gcc-8.

DanGrayson replied:
Now that I've rewritten the macro for you, you can test for openmp here and ignore OPENMP_LDFLAGS

mikestillman commented:
I just tried using
```
	CC="clang -I/usr/local/opt/libomp/include -Xpreprocessor -fopenmp" \
	CXX="clang++ -I/usr/local/opt/libomp/include -Xpreprocessor -fopenmp" \
	LDFLAGS="-L/usr/local/opt/libomp/lib" \
	LDLIBS="-lomp" 
```
in my configure script for M2, for apple clang, and the test for open mp failed in the call to configure.

DanGrayson commented:
Is the problem that clang wants `g++ -fopenmp` instead of `g++ --openmp`?

mikestillman commented:
clang-6 or clang-7 (i.e. not from apple, but from llvm) want -fopenmp.

However, apple clang unfortunately, although it sort of includes openmp support, didn't really include it.  So one needs to do -Xpreprocessor -fopenmp, and also link with -lomp (which one needs to install using brew).

I have got both of these cases to work with open mp, just not through the configure system we have. (although, with the OPENMP_CXXFLAGS, OPENMP_LDFLAGS that I kind of hacked in, I can get it to work with both.

g++-8 works out of the box.

DanGrayson commented:
How do I install openmp with brew?

mikestillman commented:
The easiest is with `brew install llvm`,I think that is what I did, this installs latest (non-apple) clang, which is easy to figure out what version it is, as opposed to the apple version.

However, for just apple clang, one can do
`brew install libomp`

mikestillman commented:
Thanks!  That would be much appreciated!

DanGrayson commented:
Try this patch -- I rewrote the macro for you:
```
diff --git a/M2/Makefile b/M2/Makefile
index 9e6f1c15a..3c5321c88 100644
--- a/M2/Makefile
+++ b/M2/Makefile
@@ -27,7 +27,7 @@ check-make:
 	@ echo GNU make version okay
 
 make-configure : check-autoconf configure
-configure : configure.ac config/files config/ax_func_accept_argtypes.m4 config/ax_prog_cc_for_build.m4 # aclocal.m4
+configure : configure.ac config/files config/ax_func_accept_argtypes.m4 config/ax_prog_cc_for_build.m4 config/openmp.m4 # aclocal.m4
 	@ set -x ; autoconf
 
 get-tools: usr-build/aclocal-path get-m4 get-pkg get-autoconf get-automake get-libtool usr-build/aclocal-path2 which-tools
diff --git a/M2/config/openmp.m4 b/M2/config/openmp.m4
new file mode 100644
index 000000000..bf04f9c98
--- /dev/null
+++ b/M2/config/openmp.m4
@@ -0,0 +1,61 @@
+# we modify this from the autoconf version to detect what to do for apple clang
+
+
+# AC_OPENMP
+# ---------
+# Check which options need to be passed to the C compiler to support OpenMP.
+# Set the OPENMP_CFLAGS / OPENMP_CXXFLAGS / OPENMP_FFLAGS variable to these
+# options.
+# The options are necessary at compile time (so the #pragmas are understood)
+# and at link time (so the appropriate library is linked with).
+# This macro takes care to not produce redundant options if $CC $CFLAGS already
+# supports OpenMP. It also is careful to not pass options to compilers that
+# misinterpret them; for example, most compilers accept "-openmp" and create
+# an output file called 'penmp' rather than activating OpenMP support.
+AC_DEFUN([AC_OPENMP],
+[
+  OPENMP_[]_AC_LANG_PREFIX[]FLAGS=
+  AC_ARG_ENABLE([openmp],
+    [AS_HELP_STRING([--disable-openmp], [do not use OpenMP])])
+  if test "$enable_openmp" != no; then
+    AC_CACHE_CHECK([for $[]_AC_CC[] option to support OpenMP],
+      [ac_cv_prog_[]_AC_LANG_ABBREV[]_openmp],
+      [AC_LINK_IFELSE([_AC_LANG_OPENMP],
+	 [ac_cv_prog_[]_AC_LANG_ABBREV[]_openmp='none needed'],
+	 [ac_cv_prog_[]_AC_LANG_ABBREV[]_openmp='unsupported'
+	  dnl Try these flags:
+	  dnl   GCC >= 4.2           -fopenmp
+	  dnl   apple clang	     -Xpreprocessor -fopenmp -lomp
+	  dnl   SunPRO C             -xopenmp
+	  dnl   Intel C              -openmp
+	  dnl   SGI C, PGI C         -mp
+	  dnl   Tru64 Compaq C       -omp
+	  dnl   IBM C (AIX, Linux)   -qsmp=omp
+          dnl   Cray CCE             -homp
+          dnl   NEC SX               -Popenmp
+          dnl   Lahey Fortran (Linux)  --openmp
+	  dnl If in this loop a compiler is passed an option that it doesn't
+	  dnl understand or that it misinterprets, the AC_LINK_IFELSE test
+	  dnl will fail (since we know that it failed without the option),
+	  dnl therefore the loop will continue searching for an option, and
+	  dnl no output file called 'penmp' or 'mp' is created.
+	  for ac_option in -fopenmp "-Xpreprocessor -fopenmp -lomp" -xopenmp -openmp -mp -omp -qsmp=omp -homp \
+                           -Popenmp --openmp; do
+	    ac_save_[]_AC_LANG_PREFIX[]FLAGS=$[]_AC_LANG_PREFIX[]FLAGS
+	    _AC_LANG_PREFIX[]FLAGS="$[]_AC_LANG_PREFIX[]FLAGS $ac_option"
+	    AC_LINK_IFELSE([_AC_LANG_OPENMP],
+	      [ac_cv_prog_[]_AC_LANG_ABBREV[]_openmp=$ac_option])
+	    _AC_LANG_PREFIX[]FLAGS=$ac_save_[]_AC_LANG_PREFIX[]FLAGS
+	    if test "$ac_cv_prog_[]_AC_LANG_ABBREV[]_openmp" != unsupported; then
+	      break
+	    fi
+	  done])])
+    case $ac_cv_prog_[]_AC_LANG_ABBREV[]_openmp in #(
+      "none needed" | unsupported)
+	;; #(
+      *)
+	OPENMP_[]_AC_LANG_PREFIX[]FLAGS=$ac_cv_prog_[]_AC_LANG_ABBREV[]_openmp ;;
+    esac
+  fi
+  AC_SUBST([OPENMP_]_AC_LANG_PREFIX[FLAGS])
+])
diff --git a/M2/configure.ac b/M2/configure.ac
index 9863526f7..923a0f98f 100644
--- a/M2/configure.ac
+++ b/M2/configure.ac
@@ -55,6 +55,7 @@ m4_include(config/ax_prog_cc_for_build.m4)
 m4_include(config/ax_func_accept_argtypes.m4)
 m4_include(config/relpaths.m4)
 m4_include(config/gtest.m4)
+m4_include(config/openmp.m4)
 
 dnl define(TO_UPPER,[translit($1, [a-z], [A-Z])])
 
```

mikestillman commented:
Thanks Dan!  I'll try this this afternoon.  Thomas, if you try it before me, check it in, and I'll try it.  Otherwise, I'll take care of that.

tom111 commented:
I tried with the rewritten macro and I can't get it to work with MacOS Mojave (the latest that came out in the end of September 2018) included clang, but I do get errors that seem to be unrelated to OpenMP.

I'm trying the following configure line:
```
../../configure \
         OPENMP_CXXFLAGS="-I/usr/local/opt/libomp/include -Xpreprocessor -fopenmp" \
         OPENMP_LDFLAGS="-L/usr/local/opt/libomp/lib -lomp" \
         LDFLAGS="-L/usr/local/opt/libxml2/lib" \
         CPPFLAGS="-I/usr/local/opt/libxml2/include/libxml2" \
         --build=x86_64-apple-darwin \
         --prefix="`pwd`/installed" \
         --disable-strip \
         --enable-download
```

Note the weird directory in CPPFLAGS. If I omit it, then one of the auto-generated headers in the d directy that contains `include <libxml/parser.h>` will trigger an error.

With this configure line I only get to errors of the following form
```
../../../../../../Macaulay2/e/x-factor.cpp:287:39: error: invalid operands to binary expression ('std::__1::ostream'
      (aka 'basic_ostream<char>') and 'Variable')
void showvar(Variable &t) { std::cout << t << std::endl; }
```

I also tried the "latestclang" from Mike's makefile, but this fails when building CSDP:
```
gmake[4]: Entering directory '/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/builds.tmp/darwin64-latestclang/libraries/csdp/build/Csdp-6.2.0/lib'
gcc -I/usr/local/opt/llvm/include -fopenmp -m64 -march=native -mtune=native -Ofast -ansi -Wall -DBIT64 -DUSEOPENMP -DSETNUMTHREADS -DUSESIGTERM -DUSEGETTIME -I../include   -c -o readprob.o readprob.c
clang: error: unsupported option '-fopenmp'
```

Finally, I tried gcc-8 which now also fails because lrslib can't find gmp.h:
```
gmake[3]: Entering directory '/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/builds.tmp/darwin64-gcc8/libraries/lrslib/build/lrslib-062'
gmake[3]: warning: -jN forced in submake: disabling jobserver mode.
gcc-8 -c -O3  -DTIMES -DSIGNALS -DGMP -o lrslib-GMP.o lrslib.c
In file included from lrslib.h:51,
                 from lrslib.c:27:
lrsgmp.h:33:10: fatal error: gmp.h: No such file or directory
 #include "gmp.h"
          ^~~~~~~
```

I'm unsure if all of these are related (with each other) and with the OpenMP changes, but it is quite weird that nothing works anymore.  

What is going on?

tom111 commented:
> I tried with the rewritten macro and I can't get it to work with MacOS Mojave (the latest that came out in the end of September 2018) included clang, but I do get errors that seem to be unrelated to OpenMP.
> 
> I'm trying the following configure line:
> 
> ```
> ../../configure \
>          OPENMP_CXXFLAGS="-I/usr/local/opt/libomp/include -Xpreprocessor -fopenmp" \
>          OPENMP_LDFLAGS="-L/usr/local/opt/libomp/lib -lomp" \
>          LDFLAGS="-L/usr/local/opt/libxml2/lib" \
>          CPPFLAGS="-I/usr/local/opt/libxml2/include/libxml2" \
>          --build=x86_64-apple-darwin \
>          --prefix="`pwd`/installed" \
>          --disable-strip \
>          --enable-download
> ```
> 
> Note the weird directory in CPPFLAGS. If I omit it, then one of the auto-generated headers in the d directy that contains `include <libxml/parser.h>` will trigger an error.
> 
> With this configure line I only get to errors of the following form
> 
> ```
> ../../../../../../Macaulay2/e/x-factor.cpp:287:39: error: invalid operands to binary expression ('std::__1::ostream'
>       (aka 'basic_ostream<char>') and 'Variable')
> void showvar(Variable &t) { std::cout << t << std::endl; }
> ```

I checked on the master branch today and it also fails with the same error. So this is unrelated with OpenMP/CSDP stuff.

DanGrayson commented:
Try changing OPENMP_CXXFLAGS to CXXFLAGS and OPENMP_LDFLAGS to LDFLAGS.  Probably setting them interferes with the macro's operation.

tom111 commented:
I reproduced the problem on the M2 master branch with this configure line:

```
../../configure LDFLAGS="-L/usr/local/opt/libxml2/lib" CPPFLAGS="-I/usr/local/opt/libxml2/include/libxml2" --enable-download
```

If I leave out the LDFLAGS and CPPFLAGS for libxml2 it fails to find parser.h at some point. This is with the Apple provided clang, as no other environment variables are configured.

DanGrayson commented:
I don't have a directory called `/usr/local/opt/libxml2/lib`.  What is that?

DanGrayson commented:
I get one if I use brew to install xml2...

tom111 commented:
Yes, it is libxml from homebrew. You can also use your local libxml from Apple. On my computer parser.h is here:
`/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/libxml2/libxml/parser.h`
For me (on MacOS Mojave), M2's build system is not able to find libxml2 so that it works without an explicit LDFLAGS and CPPFLAGS entry.

DanGrayson commented:
I'll upgrade and see if I can reproduce the problem ...

mikestillman commented:
I just installed brew on macosx (high sierra), and it gave me the following messages, when I did `brew install libomp`

```
On Apple Clang, you need to add several options to use OpenMP's front end
instead of the standard driver option. This usually looks like
  -Xpreprocessor -fopenmp -lomp

You might need to make sure the lib and include directories are discoverable
if /usr/local is not searched:

  -L/usr/local/opt/libomp/lib -I/usr/local/opt/libomp/include

For CMake, the following flags will cause the OpenMP::OpenMP_CXX target to
be set up correctly:
  -DOpenMP_CXX_FLAGS="-Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include" -DOpenMP_CXX_LIB_NAMES="omp" -DOpenMP_omp_LIBRARY=/usr/local/opt/libomp/lib/libomp.dylib
```

DanGrayson commented:
Okay, I've just upgraded to mac os x 10.14, and the libxml2 files are in `/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/libxml2/libxml/`.  I've started a Macaulay2 build, and will make M2 work with the libxml2 provided by the OS.

DanGrayson commented:
Okay, I've added the fix to the INSTALL file in commit 7796158b51b4fa2fc80088be06571b85a6269c81 .  Try using it instead of the libxml2 from brew.

tom111 commented:
Thanks for the hint. (With every release of MacOS there are more things one needs to set up before having a reasonably unix like environment.) 

In any case, for me it now works with OpenMP and with all three compilers on MacOS: Apple provided clang, brew clang-7, gcc-8.  I used the configure lines from Mike's makefile (I updated the one for "latestclang" so that it works now -- Mike, I hope it is OK, please check)

I also tried it on Linux with the linux-opt64 target. This also works.

For me it looks like you could merge this pull request now.

DanGrayson commented:
Okay, I'll take a look at it.

By the way, one source of confusion is that we recommend installing pkg-config using homebrew, and it assumes that the standard Mac OS X libxml-2.0 can be found under /usr/include -- so we all we have to do is put it there to make the statements of pkg-config true.  Our configure script depends on that information for knowing how to include the libxml files.

DanGrayson commented on M2/BUILD/mike/Makefile:
```diff
@@ -17,6 +17,79 @@ cmake-build: always
+		CC="/usr/local/opt/llvm/bin/clang" \
+	        CXX="/usr/local/opt/llvm/bin/clang++" \
+                LDFLAGS="-L/usr/local/opt/llvm/lib -Wl,-rpath,/usr/local/opt/llvm/lib" \
+		OPENMP_CXXFLAGS="-I/usr/local/opt/llvm/include -fopenmp" \
```
We don't document any use of variables OPENMP_CXXFLAGS and OPENMP_LDFLAGS.  I bet CXXFLAGS and LDFLAGS would work just as well.  Could you try?

DanGrayson commented on M2/INSTALL:
```diff
@@ -182,10 +182,24 @@ various modern operating systems:
+	brew install --build-bottle bdw-gc libatomic_ops gdbm glpk yasm make cmake
       You may omit the option "--build-bottle" if you will not be distributing your Macaulay2
       binaries to other people with CPU's that may not have the same capabilities as yours.
+    Starting with M2 version 1.12.0.1, (i.e. 1.13), the apple clang c++ compiler has some annoying
```
We should try to make this work with the native compilers provided with Mac OS X.

tom111 commented:
Hi, I pushed two new commits.

1) Update INSTALL file: I removed all the warnings about Apple's native compiler since it just works now.
2) Changed Mike's Makefile to contain the new simplified configure lines and tested the targets `darwin64-appleclang` and `darwin64-latestclang` there.

I also confirmed that after following the MacOS Instructions in INSTALL the simple configure line `../../configure --enable-download` now suffices to compile Macaulay2 with the Apple provided clang.

mikestillman commented:
@tom111, I'm testing it now, on 3 compilers: appleclang, clang7 (from homebrew llvm) and gcc8.

What else do we need to do before we are done on this?  We need to make sure the license for csdp is in the doc, like we do for all libraries.  I'll look into this.

DanGrayson commented on M2/INSTALL:
```diff
@@ -174,7 +174,7 @@ various modern operating systems:
         this command:
 	  sudo installer -pkg macOS_SDK_headers_for_macOS_10.14.pkg -target /
-	That file is found in the directory
+	The pkg file is found in the directory
```
"pkg" is not an English word.

tom111 replied:
Hmm, OK, but the formulation with "that file" confused me.

DanGrayson commented on M2/configure.ac:
```diff
@@ -741,6 +742,28 @@ AC_COMPILE_IFELSE([AC_LANG_SOURCE([
+dnl Check for openmp and fail if it cannot be found.
+dnl This is required for building csdp.
+AC_MSG_CHECKING(whether OpenMP is supported by the compiler)
+if test ! "$OPENMP_LDFLAGS"
```
Don't make the test conditional.  You should always test.

tom111 replied:
I agree.

DanGrayson commented on M2/include/configuration.in:
```diff
@@ -81,6 +81,8 @@ NODENAME=@NODENAME@
 OBJDUMP=@OBJDUMP@
 OBJEXT=@OBJEXT@
 OPTIMIZE=@OPTIMIZE@
+OPENMP_CXXFLAGS=@OPENMP_CXXFLAGS@
```
Let's not use variables with these two names, as they are not needed.

tom111 replied:
Then I don't understand the logic of the (rewritten) AC_OPENMP macro. Isn't the effect of the macro to populate these variables in line 57:
`OPENMP_[]_AC_LANG_PREFIX[]FLAGS=$ac_cv_prog_[]_AC_LANG_ABBREV[]_openmp ;;`
?

DanGrayson replied:
Oops, good point!  But then let's use them closer to the use of the macro, by changing
```
AC_SUBST(OPENMP_CXXFLAGS,"$OPENMP_CXXFLAGS")
AC_SUBST(OPENMP_LDFLAGS,"")
```
to
```
CXXFLAGS="$CXXFLAGS $OPENMP_CXXFLAGS"
LDFLAGS="$LDFLAGS $OPENMP_CXXFLAGS"
```
It's good to keep mysteries localized.

tom111 replied:
I think it does not work like this. I'm pushing a commit in a minute where I implemented this solution, but then compiling gc fails during the linking stage. I think adding -fopenmp to the LDFLAGS is not OK with Apple clang.

DanGrayson commented on M2/include/config.Makefile.in:
```diff
@@ -37,6 +37,8 @@ CXXFLAGS = @CXXFLAGS@
 FCFLAGS = @FCFLAGS@
 LDFLAGS = @LDFLAGS@
 LDFLAGS_FOR_BUILD = @LDFLAGS_FOR_BUILD@
+OPENMP_CXXFLAGS = @OPENMP_CXXFLAGS@
```
Let's not use variables with these two names, as they are not needed.

mikestillman commented:
We are not using `@OPENCXX_FLAGS@` in any Makefile.in, is that right?  SO that is why we don't need to use AC_SUBST.  But $OPENMP_CXXFLAGS is being used in, e.g. csdp/Makefile.in.  So our makefiles insure that OPENMP_CXXFLAGS has the correct value at that point?  

I recall when I first tried doing this (earlier in this pull request), that I had difficulty getting that to happen.  It was probably just my ignorance about ./configure, etc...

tom111 commented:
My new commits make the test for OpenMP unconditional and localize the usage of OPENMP_CXXFLAGS to just being appended to CXXFLAGS and LDLAGS immediately after the macro runs.
Now GC fails to link and I don't know how to continue:
```
make: Entering directory '/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/libraries/gc/build/gc-7.6.2'
+ gmake -j1 prefix=/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/usr-host
gmake[1]: Entering directory '/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/libraries/gc/build/gc-7.6.2'
gmake[2]: Entering directory '/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/libraries/gc/build/gc-7.6.2'
/bin/sh ./libtool  --tag=CC   --mode=link gcc   -fexceptions -Wall -Wextra -Wpedantic -Wno-long-long -std=gnu11 -g3 -O2 -fno-strict-aliasing  -Wl,-single_module -version-info 4:1:2 -no-undefined -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/usr-host/lib  -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/submodules/memtailor/.libs  -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/submodules/mathic/.libs  -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/submodules/mathicgb/.libs -g3 -Xpreprocessor -fopenmp -lomp -headerpad_max_install_names -o libgc.la -rpath /Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/usr-host/lib allchblk.lo alloc.lo blacklst.lo checksums.lo dbg_mlc.lo dyn_load.lo finalize.lo gc_dlopen.lo gcj_mlc.lo headers.lo mach_dep.lo malloc.lo mallocx.lo mark.lo mark_rts.lo misc.lo new_hblk.lo obj_map.lo os_dep.lo pcr_interface.lo ptr_chck.lo real_malloc.lo reclaim.lo specific.lo stubborn.lo typd_mlc.lo  pthread_start.lo pthread_support.lo thread_local_alloc.lo darwin_stop_world.lo   fnlz_mlc.lo  -lpthread  -L/usr/local/Cellar/libatomic_ops/7.6.6/lib -latomic_ops
libtool: link: gcc -dynamiclib  -o .libs/libgc.2.dylib  .libs/allchblk.o .libs/alloc.o .libs/blacklst.o .libs/checksums.o .libs/dbg_mlc.o .libs/dyn_load.o .libs/finalize.o .libs/gc_dlopen.o .libs/gcj_mlc.o .libs/headers.o .libs/mach_dep.o .libs/malloc.o .libs/mallocx.o .libs/mark.o .libs/mark_rts.o .libs/misc.o .libs/new_hblk.o .libs/obj_map.o .libs/os_dep.o .libs/pcr_interface.o .libs/ptr_chck.o .libs/real_malloc.o .libs/reclaim.o .libs/specific.o .libs/stubborn.o .libs/typd_mlc.o .libs/pthread_start.o .libs/pthread_support.o .libs/thread_local_alloc.o .libs/darwin_stop_world.o .libs/fnlz_mlc.o   -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/usr-host/lib -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/submodules/memtailor/.libs -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/submodules/mathic/.libs -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/submodules/mathicgb/.libs -lomp -lpthread -L/usr/local/Cellar/libatomic_ops/7.6.6/lib -latomic_ops  -g3 -O2 -Wl,-single_module -g3 -fopenmp   -fopenmp -install_name  /Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/usr-host/lib/libgc.2.dylib -compatibility_version 5 -current_version 5.1 -Wl,-single_module
clang: error: unsupported option '-fopenmp'
```

tom111 commented:
>  But $OPENMP_CXXFLAGS is being used in, e.g. csdp/Makefile.in. So our makefiles insure that OPENMP_CXXFLAGS has the correct value at that point?

I think it is not necessary to use these variables there if we can just append their values to CXXFLAGS and LDFLAGS. In fact, I tried that CSDP compiles just fine without them after commit 
85b5cd6.
Just GC fails now.

DanGrayson commented:
> make: Entering directory '/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/libraries/gc/build/gc-7.6.2' + gmake -j1 prefix=/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/usr-host gmake[1]: Entering directory '/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/libraries/gc/build/gc-7.6.2' gmake[2]: Entering directory '/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/libraries/gc/build/gc-7.6.2' /bin/sh ./libtool --tag=CC --mode=link gcc -fexceptions -Wall -Wextra -Wpedantic -Wno-long-long -std=gnu11 -g3 -O2 -fno-strict-aliasing -Wl,-single_module -version-info 4:1:2 -no-undefined -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/usr-host/lib -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/submodules/memtailor/.libs -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/submodules/mathic/.libs -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/submodules/mathicgb/.libs -g3 -Xpreprocessor -fopenmp -lomp -headerpad_max_install_names -o libgc.la -rpath /Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/usr-host/lib allchblk.lo alloc.lo blacklst.lo checksums.lo dbg_mlc.lo dyn_load.lo finalize.lo gc_dlopen.lo gcj_mlc.lo headers.lo mach_dep.lo malloc.lo mallocx.lo mark.lo mark_rts.lo misc.lo new_hblk.lo obj_map.lo os_dep.lo pcr_interface.lo ptr_chck.lo real_malloc.lo reclaim.lo specific.lo stubborn.lo typd_mlc.lo pthread_start.lo pthread_support.lo thread_local_alloc.lo darwin_stop_world.lo fnlz_mlc.lo -lpthread -L/usr/local/Cellar/libatomic_ops/7.6.6/lib -latomic_ops libtool: link: gcc -dynamiclib -o .libs/libgc.2.dylib .libs/allchblk.o .libs/alloc.o .libs/blacklst.o .libs/checksums.o .libs/dbg_mlc.o .libs/dyn_load.o .libs/finalize.o .libs/gc_dlopen.o .libs/gcj_mlc.o .libs/headers.o .libs/mach_dep.o .libs/malloc.o .libs/mallocx.o .libs/mark.o .libs/mark_rts.o .libs/misc.o .libs/new_hblk.o .libs/obj_map.o .libs/os_dep.o .libs/pcr_interface.o .libs/ptr_chck.o .libs/real_malloc.o .libs/reclaim.o .libs/specific.o .libs/stubborn.o .libs/typd_mlc.o .libs/pthread_start.o .libs/pthread_support.o .libs/thread_local_alloc.o .libs/darwin_stop_world.o .libs/fnlz_mlc.o -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/usr-host/lib -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/submodules/memtailor/.libs -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/submodules/mathic/.libs -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/submodules/mathicgb/.libs -lomp -lpthread -L/usr/local/Cellar/libatomic_ops/7.6.6/lib -latomic_ops -g3 -O2 -Wl,-single_module -g3 -fopenmp -fopenmp -install_name /Users/tom/M2-toms-fork-on-github/M2/BUILD/tom/usr-host/lib/libgc.2.dylib -compatibility_version 5 -current_version 5.1 -Wl,-single_module clang: error: unsupported option '-fopenmp'

That looks like a libtool bug.  It's mangling the command line options.

DanGrayson commented:
Yes, it's a libtool bug.  But I suspect you're plugging "-Xpreprocessor -fopenmp" into LDFLAGS.  If so, can you try without it?  Or is it needed?

If not, we'll have to go back to your original idea of propagating the needed flags in a separate configure variable.

tom111 commented:
> Yes, it's a libtool bug. But I suspect you're plugging "-Xpreprocessor -fopenmp" into LDFLAGS. If so, can you try without it? Or is it needed?

I tried to remove them, but the LDFLAGS are not the problem. Libtool is inserting `-g3 -O2 -fopenmp -g3   -fopenmp` somewhere near the end of the commandline and I don't know where it comes from. It does not change if LDFLAGS only contains "-lomp"

I think libtool tries to be smart about `-fopenmp`. It has some special handling for it which fails here.

DanGrayson commented:
I agree.  We'll just have to prevent libtool from seeing those two flags.

mikestillman commented:
Hi @tom111 , where do we stand here?  When I posted the message 3 days ago, before the latest changes, it worked on all 3 mac c++ compilers: apple clang, latest clang, gcc8.  

The current problem is that libtool is incompatible with aplpe clang openmp, as it expects a certain behavior for how to invoke openmp.

We should not be passing openmp to our libraries that don't need it.  Maybe we should use a different name for OPENMP_CXXFLAGS, so it doesn't conflict with the detection module?

tom111 commented:
Hi @mikestillman. Yes, somebody needs to revert the commits and restore the state where everything was working. Dan did not like that particular solution until we uncovered this libtool bug. 

Going without these flags is not the solution as we then lose support for OpenMP in normaliz which I don't like. If I had time I would revert to the state where we are passing OPENMP_CXX flags to the CSDP makefile. 

Then I also wanted to investigate the C vs. C++ issue with csdp.  I'm not sure if it actually uses OpenMP the way we do it.

I have very little time this week, but I'll be in Providence next week.  Feel free to go ahead and revert or I can do it as soon as I have some time.

tom111 commented:
Hi again. Status update: I've rebased to Macaulay2 main master and made the check for OpenMP in configure.ac unconditional again as this was one of Dan's original points.

Now the status here is:

- Apple Clang: Compiles fine with just `../../configure --enable-download` but I believe normaliz does not use OpenMP, at least `otool -L csdp` shows no dependency on libomp.
- Clang-7 from Homebrew: Compiles fine `CC="/usr/local/opt/llvm/bin/clang" 
                CXX="/usr/local/opt/llvm/bin/clang++" 
                LDFLAGS="-L/usr/local/opt/llvm/lib -Wl,-rpath,/usr/local/opt/llvm/lib" 
                CPPFLAGS="-I/usr/local/opt/llvm/include"
                --enable-download` and normaliz and csdp are both linked against libomp.  So everything is really fine here.
- GCC-8 with: `CC=gcc-8 CXX=g++-8 CPP=cpp-8 LD=gcc-8 ../../configure --enable-download` fails to link normaliz now. This issue seems to be unrelated to OpenMP, I will try to reproduce it on master and then report a new issue:
```
libtool: link: g++-8 -fopenmp -isystem /Users/tom/M2-toms-fork-on-github/M2/BUILD/mike/builds.tmp/darwin64-gcc8/../../../../include -isystem /Users/tom/M2-toms-fork-on-github/M2/BUILD/mike/builds.tmp/darwin64-gcc8/include -isystem /Users/tom/M2-toms-fork-on-github/M2/BUILD/mike/builds.tmp/darwin64-gcc8/usr-host/include -isystem /Users/tom/M2-toms-fork-on-github/M2/BUILD/mike/builds.tmp/darwin64-gcc8/usr-host/include/cdd -I/Users/tom/M2-toms-fork-on-github/M2/BUILD/mike/builds.tmp/darwin64-gcc8/../../../../submodules/memtailor/src -I/Users/tom/M2-toms-fork-on-github/M2/BUILD/mike/builds.tmp/darwin64-gcc8/../../../../submodules/mathic/src -I/Users/tom/M2-toms-fork-on-github/M2/BUILD/mike/builds.tmp/darwin64-gcc8/../../../../submodules/mathicgb/src -I/Users/tom/M2-toms-fork-on-github/M2/BUILD/mike/builds.tmp/darwin64-gcc8/../../../../submodules/fflas_ffpack/src -isystem /usr/include/libxml2 -isystem /usr/local/Cellar/libatomic_ops/7.6.6/include -DNDEBUG -Wall -O3 -Wno-unknown-pragmas -std=c++11 -I .. -I . -Wno-unknown-pragmas -DNO_OPENMP -g3 -headerpad_max_install_names -o normaliz normaliz.o -Wl,-bind_at_load  -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/mike/builds.tmp/darwin64-gcc8/usr-host/lib -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/mike/builds.tmp/darwin64-gcc8/submodules/memtailor/.libs -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/mike/builds.tmp/darwin64-gcc8/submodules/mathic/.libs -L/Users/tom/M2-toms-fork-on-github/M2/BUILD/mike/builds.tmp/darwin64-gcc8/submodules/mathicgb/.libs ./.libs/libnormaliz.a -lflint -lmpfr -lgc -lgmpxx -lgmp -lpthread -fopenmp
ld: warning: directory not found for option '-L/Users/tom/M2-toms-fork-on-github/M2/BUILD/mike/builds.tmp/darwin64-gcc8/submodules/memtailor/.libs'
ld: warning: directory not found for option '-L/Users/tom/M2-toms-fork-on-github/M2/BUILD/mike/builds.tmp/darwin64-gcc8/submodules/mathic/.libs'
ld: warning: directory not found for option '-L/Users/tom/M2-toms-fork-on-github/M2/BUILD/mike/builds.tmp/darwin64-gcc8/submodules/mathicgb/.libs'
Undefined symbols for architecture x86_64:
  "___gmpn_com", referenced from:
      __fmpz_vec_get_fft in libflint.a(fmpz_vec_get_fft.o)
      __fmpz_vec_set_fft in libflint.a(fmpz_vec_set_fft.o)
      _fft_mulmod_2expp1 in libflint.a(fft_mulmod_2expp1.o)
      _fmpz_bit_pack in libflint.a(fmpz_bit_pack.o)
      _fmpz_bit_unpack in libflint.a(fmpz_bit_unpack.o)
      _fft_butterfly_twiddle in libflint.a(fft_fft_mfa_truncate_sqrt2.o)
      _fft_butterfly_sqrt2 in libflint.a(fft_fft_truncate_sqrt2.o)
      ...
     (maybe you meant: ___gmpn_com_n)
ld: symbol(s) not found for architecture x86_64
```

mikestillman commented:
normaliz uses this in configure.ac:

```
AC_ARG_ENABLE([openmp],
    [AS_HELP_STRING([--enable-openmp@<:@=ARG@:>@],
      [enable parallelization with OpenMP @<:@default=check@:>@])],
    [:],
    [enable_openmp=check])
AS_IF([test "x$enable_openmp" != xno],
  [AX_OPENMP([have_openmp=yes],
	     [have_openmp=no])])
```

They provide an m4/ax_openmp.m4, which we probably should patch: i.e. use the one Dan just wrote (I don't think we are doing that).  Does this seem correct?  I'm not sure if there version sets any other variables that normaliz would need (but I haven't really looked yet).

tom111 commented:
> normaliz uses this in configure.ac:

Oh, yes, of course.  This ax_openmp macro that normaliz uses does not check for the option `-Xpreprocessor -fopenmp`, so it will never detect openmp support on apple clang.  Let's replace it by Dan's version or patch it to also check for `-Xpreprocessor -fopenmp`. I could work on it tomorrow.

mikestillman commented:
I don't think I'll have time today to do it.  One option is we patch the configure file, removing the test, and pass in the correct CXX_FLAGS, CC_FLAGS, LDLIBS, LD_FLAGS, etc. to normaliz since we have already checked that (we might be doing that already?)

tom111 commented:
Hmm, Normaliz uses libtool. I think we are screwed. I don't know how to compile normaliz with OpenMP on Apple clang without starting to debug libtool. Shall we just give up and put a note in the INSTALL file that if you want openMP support in normaliz (which does speed it up a lot!), then you have to use homebrew clang or gcc?

mikestillman commented:
One other option is that normaliz also has a CMakeLists.txt file, so maybe they support cmake.  We could try building it with cmake.

That said, I agree with you about suggesting latest clang, or gcc, and punting on all of this.  When we create binaries of M2, we should be careful to not use apple clang, too.  @DanGrayson is this OK?

tom111 commented:
> Why do that?

Close to linear speedup in the number of processors until about 5k-1M processors, depending on the problem.  We can build it without OpenMP but this limits the usefulness of the normaliz package a lot.

DanGrayson commented:
Re: " I don't know how to compile normaliz with OpenMP"

I thought the plan was to compile it without openmp.

On Wed, Nov 7, 2018 at 3:40 PM Mike Stillman <notifications@github.com>
wrote:

> One other option is that normaliz also has a CMakeLists.txt file, so maybe
> they support cmake. We could try building it with cmake.
>
> That said, I agree with you about suggesting latest clang, or gcc, and
> punting on all of this. When we create binaries of M2, we should be careful
> to not use apple clang, too. @DanGrayson <https://github.com/DanGrayson>
> is this OK?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Macaulay2/M2/pull/850#issuecomment-436644745>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAqvRMLb014YuLBpxRq3hwcdFP3lNioxks5usvDtgaJpZM4Wha5D>
> .
>

tom111 commented:
For me it is OK to not support OpenMP enabled normaliz with Apple Clang.  Before we started this project here, this was the state for all compilers, so there is some improvement already. 

It seems there is only Issue #881 left and this might be unrelated. Or is there anything else?

DanGrayson commented on M2/INSTALL:
```diff
@@ -188,10 +188,11 @@ various modern operating systems:
-	brew install --build-bottle pkg-config autoconf automake gnu-tar libtool boost libmpc xz wget
-	brew install --build-bottle bdw-gc libatomic_ops gdbm glpk yasm make
+	brew install --build-bottle pkg-config autoconf automake gnu-tar libtool boost libmpc xz wget libomp
+	brew install --build-bottle bdw-gc libatomic_ops gdbm glpk yasm make cmake
```
Do we need cmake now??

mikestillman replied:
I added that in.  The answer is no, we do not need cmake now.  We can remove it.

DanGrayson commented on M2/configure.ac:
```diff
@@ -741,6 +742,18 @@ AC_COMPILE_IFELSE([AC_LANG_SOURCE([
+   then AC_MSG_ERROR([OpenMP does not work, use a compiler that supports OpenMP])
+fi
+AC_SUBST(OPENMP_CXXFLAGS,"$OPENMP_CXXFLAGS")
+AC_SUBST(OPENMP_LDFLAGS,"")
```
What's the purpose of this?

mikestillman replied:
What is the purpose of what?  I don't understand your question.  Do you mean just the AC_SUBST(OPENMP_LDFLAGS, "") line, or the whole thing (which we have benn discussing the need for in the last week).

DanGrayson replied:
Yes, what is the purpose of adding OPENMP_LDFLAGS if its value is always empty?

tom111 commented:
Good morning everybody. I've pushed two new commits. One removes cmake from the suggested brewline, the other removes the unused variable OPENMP_LDFLAGS.

DanGrayson commented:
There's a problem: the Csdp makefile has this:
```
#                                                                                                                                                          
# Install the executables in /usr/local/bin.                                                                                                               
#                                                                                                                                                          

install:
        cp -f solver/csdp /usr/local/bin
        cp -f theta/theta /usr/local/bin
        cp -f theta/graphtoprob /usr/local/bin
	cp -f theta/complement /usr/local/bin
        cp -f theta/rand_graph /usr/local/bin
```
The result is this:
```
+ /usr/bin/make -j1 prefix=/mike-raid/home/dan/src/M2/overnight/M2/BUILD/dan/builds.tmp/habanero-master/usr-host install
make[1]: Entering directory '/mike-raid/home/dan/src/M2/overnight/M2/BUILD/dan/builds.tmp/habanero-master/libraries/csdp/build/Csdp-6.2.0'
cp -f solver/csdp /usr/local/bin
cp: cannot create regular file '/usr/local/bin/csdp': Permission denied
```

Can you change it from /usr/local/bin to $(prefix)/bin?

(Do git pull first, I fixed some other things.)

DanGrayson commented:
Never mind, I fixed it.

dimpase commented:
by the way, Sage uses an autotoolized csdp (which is a bit behind csdp master atm)
https://github.com/dimpase/csdp.git
which might come very handy for OSX etc...

tom111 commented:
I will try to reproduce this, but do you know how it fails? The code that you have quoted does not appear in the package and it depends on `tol`. What is the value of `tol` when you run it?

tom111 commented:
I found another issue: On 32 Bit machines CSDP fails to compile because the Makefile explicitly depends on it being a 64 Bit machine.  I'm workig on this.

tom111 commented:
> I found another issue: On 32 Bit machines CSDP fails to compile because the Makefile explicitly depends on it being a 64 Bit machine. I'm workig on this.

Actually I don't know how to do that properly.

I was thinking to patch the CSDP Makefile and replace "-m64" and "-DBIT64" by the value of some variable that configure defines for me, depending on 32/64 Bit environment. Is there such a variable or is there some other code that explicitly branches on this?

DanGrayson commented:
I stopped building M2 on 32 bit machines long ago...

DanGrayson commented:
You should be able to just remove the options that select 64 bit mode.

tom111 commented:
How do I detect that I'm in 32 Bit mode?

tom111 commented:
> This test failed under Ubuntu overnight:
> 
> ```
> debug needsPackage "SOS"
>     equal = (a,b) -> (
>         if a===null then return false;
>         d := if abs(b)<1 then abs(a-b) else abs(a-b)/abs(b);
>         << "d = " << d << "   tol = " << tol << endl;
>         return d < tol;
>         );
> R = RR[x,y];
> f = (x-pi*y)^2 + x^2 + (y-4)^2;
> (bound,sol) = lowerBound(f, Solver=>"CSDP")
> equal(bound,16*pi^2/(2+pi^2))
> ```

We tried to reproduce this on a 64 Bit Ubuntu with M2 build from the master branch. All tests pass and checkSolver returns no errors. I don't know how to further debug this.

dimpase commented:
> > I found another issue: On 32 Bit machines CSDP fails to compile because the Makefile explicitly depends on it being a 64 Bit machine. I'm workig on this.
> 
> Actually I don't know how to do that properly.
> 

Why don't you let autotools take care of it, as in https://github.com/dimpase/csdp ?

tom111 commented:
> > > I found another issue: On 32 Bit machines CSDP fails to compile because the Makefile explicitly depends on it being a 64 Bit machine. I'm workig on this.
> > 
> > 
> > Actually I don't know how to do that properly.
> 
> Why don't you let autotools take care of it, as in https://github.com/dimpase/csdp ?
> The latter builds and runs on 64 and 32 bit Linux, OSX, Windows (cygwin)...

(I'm also at ICERM, let's talk about it in person) Short version:
- It's a lot of work to switch to your fork of CSDP inside Macaulay2 now that we have it working 99%.
- Your version does not set -DBIT64 at all, so you will compile the 32 Bit Fortran conversion, no matter if the system is 64 Bit or not.

DanGrayson commented:
What I cited is actually case "t1" from this code of yours:
```
-- check lowerBound
checkLowerBound = (solver,applyTest) -> (
    tol := 0.001;
    local x; x= symbol x;
    local y; y= symbol y;
    local z; z= symbol z;
    local mult;
    equal := (a,b) -> (
        if a===null then return false;
        d := if abs(b)<1 then abs(a-b) else abs(a-b)/abs(b);
        return d < tol;
        );
    cmp := (f,h,bound,mon,Q,mult) -> (
        if Q===null then return false;
        d := f - bound + (h*mult - transpose mon * Q * mon)_(0,0);
        return isZero(MedPrecision, d);
        );

    --------------UNCONSTRAINED1--------------
    t0:= if applyTest(0) then(
        R := QQ[x];
        f := (x-1)^2 + (x+3)^2;
        (bound,sol) := lowerBound(f, Solver=>solver);
        equal(bound,8)
        );

    t1:= if applyTest(1) then(
        R = RR[x,y];
        f = (x-pi*y)^2 + x^2 + (y-4)^2;
        (bound,sol) = lowerBound(f, Solver=>solver);
        equal(bound,16*pi^2/(2+pi^2))
        );

```

DanGrayson commented:
Re: "How do I detect that I'm in 32 Bit mode?"

I think it's up to the builder of Macaulay2 to specify the desired mode with compiler options, so you can just ignore the difference.  But that means that the option line in the csdp makefile should have those architecture options removed, i.e. change
```
export CFLAGS += -m64 -march=native -mtune=native -Ofast -ansi -Wall -DBIT64 -DUSEOPENMP -DSETNUMTHREADS -DUSESIGTERM -DUSEGETTIME -I../include
```
to
```
export CFLAGS +=  -ansi -Wall -DUSEOPENMP -DSETNUMTHREADS -DUSESIGTERM -DUSEGETTIME -I../include
```
